portal/src/main/webapp/js/src/patient-view/util/clinicalAttributesUtil.js

```diff
@@ -0,0 +1,162 @@
+    if (caseType !== null) {
+      var loc;
+
+      derivedClinicalAttributes.DERIVED_NORMALIZED_CASE_TYPE = normalizedCaseType(caseType);
+
+      // TODO: DERIVED_SAMPLE_LOCATION should probably be a clinical attribute.
+      if (derivedClinicalAttributes.DERIVED_NORMALIZED_CASE_TYPE === 'metastasis') {
+          loc = getFirstKeyFound(clinicalData, ['METASTATIC_SITE', 'TUMOR_SITE']);
+      } else if (derivedClinicalAttributes.DERIVED_NORMALIZED_CASE_TYPE === 'primary') {
+          loc = getFirstKeyFound(clinicalData, ['PRIMARY_SITE', 'TUMOR_SITE']);
+      } else {
+          loc = getFirstKeyFound(clinicalData, ['TUMOR_SITE']);
+      }
+      if (loc !== null) {
+        derivedClinicalAttributes.DERIVED_SAMPLE_LOCATION = loc;
+      }
+    }
+
+    return derivedClinicalAttributes;
+  }
+
+  /**
+   * Run both clean and derive on the clinicalData.
+   */
+  function cleanAndDerive(clinicalData) {
+    return derive(clean(clinicalData));
+  }
+
+  /**
+   * Return string of spans representing the clinical attributes. The spans
+   * have been made specifically to add clinical attribute information as
+   * attributes to allow for easy styling with CSS.
+   * @param {object} clinicalData   - key/value pairs of clinical data
+   * @param {string} cancerStudyId  - short name of cancer study
+   * @param {string} extraClassName - extra class name to add e.g. patient-attribute
+   */
+  function getSpans(clinicalData, cancerStudyId, extraClassName) {
+    var spans = '';
+    var clinicalAttributesCleanDerived = cleanAndDerive(clinicalData);
+    var spanTemplate = '<span class="{0} clinical-attribute" attr-id="{1}" attr-value="{2}" study="{3}">{2}</span>';
+
+    for (var key in clinicalAttributesCleanDerived) {
+      var value = clinicalAttributesCleanDerived[key];
+      spans += spanTemplate.format(extraClassName, key, value, cancerStudyId);
+    }
+    
+    return spans;
+  }
+
+  /*
+   * Add .first-order class to all elements with the lowest order attribute.
+   * This way the first element can be styled in a different manner. If flex
+   * order attributes were working properly in CSS, one would be able to say
+   * .clinical-attribute:first, this is unfortunately not the case, therefore
+   * this hack is required. See clinical-attributes.css to see how this is
+   * used.
+   */
+  function addFirstOrderClass() {
+    $(".sample-record-inline, #more-patient-info").each(function(x) {
+      var orderSortedAttributes = _.sortBy($(this).find("a > .clinical-attribute"), function(y) {
+        var order = parseInt($(y).css("order"));
+        if (isNaN(order)) {
+          console.log("Warning: No order attribute found in .clinical-attribute.")
+        }
+        return order;
+      });
+      $(orderSortedAttributes[0]).addClass("first-order");
+    });
+  }
+
+
+  return {
+    cleanAndDerive: cleanAndDerive,
+    getSpans: getSpans,
+    addFirstOrderClass: addFirstOrderClass
+  };
+})();
```

Comment:
Since you already wrap them into a singleton, could you pass in getFirstKeyFound as a parameter?